### PR TITLE
Fix e2e testing logic

### PR DESCRIPTION
**Pull request by upodroid**
I cleaned up some logic in the kubetest2-kops deployer. Some flags are automatically enabled when testing on GCE.

### CODE REVIEW

**Comment by upodroid** [comment]
Got a successful test run by running the following config:

```
kubetest2 kops \
  -v 2 \
  --up --down \
  --cluster-name arm64.k8s.local \
  --cloud-provider=gce \
  --admin-access="0.0.0.0/0" \
  --gcp-project=borg-env \
  --target-build-arch="linux/arm64" \
  --create-args="--image='ubuntu-os-cloud/ubuntu-2204-jammy-arm64-v20230908' --channel=alpha --zones=us-central1-a --node-count=2 --set=spec.cloudControllerManager.image=gcr.io/borg-env/cloud-controller-manager:918af8f9-20230907T125911" \
  --kops-version-marker=https://storage.googleapis.com/kops-ci/bin/latest-ci-updown-green.txt \
  --kubernetes-version=https://storage.googleapis.com/k8s-release-dev/ci/latest.txt \
  --kubernetes-feature-gates=GRPCContainerProbe,LegacyServiceAccountTokenTracking \
  --test=kops \
  -- \
  --ginkgo-args="--debug" \
  --test-args="-test.timeout=70m -num-nodes=0" \
  --test-package-bucket=k8s-release-dev \
  --test-package-dir=ci \
  --test-package-marker=latest.txt \
  --focus-regex="\[Conformance\]" \
  --skip-regex="\[Slow\]|\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\]" \
  --parallel=30
  
 # get the kubeconfig
 kops export kubeconfig --all --admin --state gs://borg-env-state-783a
 
 #patch the ccm to run
 kubectl -n kube-system patch ds cloud-controller-manager --type=json -p='[{"op": "remove", "path": "/spec/template/spec/containers/0/command"}]'
```

File: tests/e2e/kubetest2-kops/deployer/up.go [hunk 1]
```diff
@@ -146,39 +147,54 @@ func (d *deployer) createCluster(zones []string, adminAccess string, yes bool) e
 		}
 	}
 	if !foundCPCount {
-		args = appendIfUnset(args, "--master-count", fmt.Sprintf("%d", d.ControlPlaneCount))
+		args = appendIfUnset(args, "--control-plane-count", fmt.Sprintf("%d", d.ControlPlaneCount))
```
Comment (hakman) [inline tests/e2e/kubetest2-kops/deployer/up.go:150]: We need to use `master` instead of `control-plane` for backwards compatibility. Please revert.

**Review by hakman — CHANGES_REQUESTED** [review]

File: tests/e2e/kubetest2-kops/deployer/up.go (hunk 1 above)
Comment (upodroid) [reply tests/e2e/kubetest2-kops/deployer/up.go]: Fixed

Comment (hakman) [reply tests/e2e/kubetest2-kops/deployer/up.go]: Thanks :)

Comment (hakman) [reply tests/e2e/kubetest2-kops/deployer/up.go]: Please do that everywhere, not just this specific flag.

Comment (upodroid) [reply tests/e2e/kubetest2-kops/deployer/up.go]: Revert the changes to all the flags

File: tests/e2e/kubetest2-kops/deployer/deployer.go [hunk 2]
```diff
@@ -108,8 +109,10 @@ func (d *deployer) Provider() string {
 func New(opts types.Options) (types.Deployer, *pflag.FlagSet) {
 	// create a deployer object and set fields that are not flag controlled
 	d := &deployer{
-		commonOptions:        opts,
-		BuildOptions:         &builder.BuildOptions{},
+		commonOptions: opts,
+		BuildOptions: &builder.BuildOptions{
+			TargetBuildArch: "linux/amd64",
+		},
```
Comment (hakman) [inline tests/e2e/kubetest2-kops/deployer/deployer.go:115]: Could you explain the need for `TargetBuildArch`?

Comment (upodroid) [reply tests/e2e/kubetest2-kops/deployer/deployer.go:115]: valid values would be linux/amd64 and linux/arm64.

I'm also thinking about potentially introducing flags to build kubernetes when testing k/k pull requests. You need that flag set to build the correct os/arch

You would have 3 situations:
- test a combination of prebuilt kubernetes and kops (ci jobs)
- test a prebuilt kops in a PR against kubernetes/kubernetes (this doesn't work yet)
- test a prebuilt kubernetes in a PR against kops (this what the --build flag does)

File: tests/e2e/kubetest2-kops/deployer/up.go (hunk 1 above)
Comment (hakman) [reply tests/e2e/kubetest2-kops/deployer/up.go]: There is still a mention of `--control-plane-size`.
Please remove the new "change this deprecated flag after the tester starts using kops.k8s.io/v1alpha3 APIs" comments. It is not correct.

File: tests/e2e/kubetest2-kops/deployer/up.go [hunk 3]
```diff
@@ -308,11 +330,19 @@ func (d *deployer) zones() ([]string, error) {
 }
 
 // appendIfUnset will append an argument and its value to args if the arg is not already present
-// This shouldn't be used for arguments that can be specified multiple times like --set
+// This shouldn't be used for arguments that can be specified multiple times except --set
 func appendIfUnset(args []string, arg, value string) []string {
+	setFlags := []string{}
 	for _, existingArg := range args {
 		existingKey := strings.Split(existingArg, "=")
```
Comment (hakman) [inline tests/e2e/kubetest2-kops/deployer/up.go:337]: There are cases where there are even more "=" in the expression.
```suggestion
		existingKey := strings.SplitN(existingArg, "=", 2)
```

File: tests/e2e/kubetest2-kops/deployer/up.go [hunk 4]
```diff
@@ -146,39 +147,60 @@ func (d *deployer) createCluster(zones []string, adminAccess string, yes bool) e
...
 	if !foundCPCount {
+		// change this deprecated flag after the tester starts using kops.k8s.io/v1alpha3 APIs
 		args = appendIfUnset(args, "--master-count", fmt.Sprintf("%d", d.ControlPlaneCount))
 	}
 
+	// change this deprecated flag after the tester starts using kops.k8s.io/v1alpha3 APIs
 	args = appendIfUnset(args, "--master-volume-size", "48")
 	args = appendIfUnset(args, "--node-count", "4")
 	args = appendIfUnset(args, "--node-volume-size", "48")
-	args = appendIfUnset(args, "--set", adminAccess)
```
Comment (hakman) [inline tests/e2e/kubetest2-kops/deployer/up.go:155]: What is the effect here?

File: tests/e2e/kubetest2-kops/deployer/deployer.go (hunk 2 above)
Comment (hakman) [reply tests/e2e/kubetest2-kops/deployer/deployer.go:115]: Could you please split this into a separate PR? Will need to discuss it on Thu. Thanks!

**Comment by hakman** [comment]
@upodroid I added a question and some requests. I think this can be merged once all these are addressed.

File: tests/e2e/kubetest2-kops/deployer/up.go (hunk 4 above)
Comment (upodroid) [reply tests/e2e/kubetest2-kops/deployer/up.go:155]: We already set that flag immediately after reading CreateArgs. The duplicate append was causing issues when I amended the `appendIfUnset` function

File: tests/e2e/kubetest2-kops/deployer/deployer.go (hunk 2 above)
Comment (upodroid) [reply tests/e2e/kubetest2-kops/deployer/deployer.go:115]: I left the BuildOptions change struct. We are detecting arm64 using the old method. I'll join the meeting on Thu to discuss this further.

Comment (upodroid) [reply tests/e2e/kubetest2-kops/deployer/deployer.go:115]: Addressing this in a separate PR.

**Comment by upodroid** [comment]
Can we merge this today please? I addressed what we discussed at the office hours

File: tests/e2e/kubetest2-kops/deployer/up.go [hunk 5]
```diff
@@ -152,21 +153,31 @@ func (d *deployer) createCluster(zones []string, adminAccess string, yes bool) e
...
+			args = appendIfUnset(args, "--node-size", "t2a-standard-2")
+		} else {
+			args = appendIfUnset(args, "--master-size", "e2-standard-2")
+			args = appendIfUnset(args, "--node-size", "e2-standard-2")
+		}
 		if d.GCPProject != "" {
 			args = appendIfUnset(args, "--project", d.GCPProject)
 		}
+		// set some sane default e2e testing behaviour on gce
+		args = appendIfUnset(args, "--gce-service-account", "default")
```
Comment (upodroid) [inline tests/e2e/kubetest2-kops/deployer/up.go:178]: The Google Service Account that prow runs with doesn't have permissions to make IAM changes on test projects. I can talk to @ameukam about it at a later date.

Sample failure: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/e2e-ci-kubernetes-e2e-cos-gce-serial-canary/1705258496510922752

File: tests/e2e/kubetest2-kops/deployer/deployer.go [hunk 6]
```diff
@@ -108,8 +109,10 @@ func (d *deployer) Provider() string {
 func New(opts types.Options) (types.Deployer, *pflag.FlagSet) {
 	// create a deployer object and set fields that are not flag controlled
 	d := &deployer{
-		commonOptions:        opts,
-		BuildOptions:         &builder.BuildOptions{},
+		commonOptions: opts,
+		BuildOptions: &builder.BuildOptions{
+			TargetBuildArch: "linux/amd64",
```
Comment (hakman) [inline tests/e2e/kubetest2-kops/deployer/deployer.go:114]:
```suggestion
```

Comment (upodroid) [reply tests/e2e/kubetest2-kops/deployer/deployer.go:114]: brb

**Review by hakman — APPROVED** [review]

**Comment by upodroid** [comment]
Pls rerun the flaking GitHub action

**Pull request merged**